SE-0400: State limitations (#2389)
apple/swift-evolution: proposals/0400-init-accessors.md

}
```

### Restrictions

* A property with an `init` accessor must be declared in the same source file 
  as the respective type declaration.
* An `init` accessor is always `fileprivate` and thus cannot be called from
  within a different source file.

> 

Author: I am still trying to make sense of the first one (it was inferred from [this](https://forums.swift.org/t/pitch-init-accessors/64881/16) post). If an `init` accessor is always confined to its file, then, logically, it cannot break a memberwise initializer generated in another file that isn’t aware of its existence.

Reviewer: I think this should just say that init accessors can only be written in the primary declaration of a type.

Author: ```suggestion
### Restrictions

A property with an `init` accessor can only be declared in the primary 
declaration of a type.

```

## Source compatibility

`init` accessors are an additive capability with new syntax; there is no impact on existing source code.

## ABI compatibility

> `init` accessors are only called from within a source file, so they are not part of the module's ABI.

Reviewer: I think the original text is correct. Even if you can only define init accessors in the primary declaration of a type, you can add an initializer in another source file that will call the init accessor:

```swift
// in FirstFile.swift

struct S {
  var _x: Int

  var x: Int {
    @storageRestrictions(initializes: _x)
    init {
      self._x = newValue
    }
    get { _x }
  }
}

// in SecondFile.swift

extension S {
  init() {
    self.x = 10
  }
}
```

Author: Thanks. We had a crash report with like the simplest example of an init accessor use across files, and I was not sure if this sentence really meant they could be `internal` because the forum topics appeared to settle on `fileprivate`. I could have just inferred it from the `inlinable` business that follows, really. But I still think this paragraph would benefit from a little more straightforwardness and accuracy (see the latest revision).

> `init` accessors are only called from within a source file, so they are not part of the module's ABI.
> An `init` accessor is *not* inferred as, but must be, inlinable if called from
> an inlinable type initializer.

Author: ```suggestion
`init` accessors can be ABI-public but are at most `internal`.
Calling an `init` accessor from an `inlinable` type initializer requires that
the `init` accessor is ABI-public.
```
